src/pybind/mgr/object_format.py

```diff
@@ -1,5 +1,133 @@
+Extending the previous example to use this exception:
+    @CLICommand("command name", perm="r")
+    Responder()
+    def create_something(self, name: str) -> Dict[str, str]:
+        try:
+           ...  # implementation
+           return {"name": name, "id": new_id}
+        except KeyError as kerr:
+           # explicitly set the return value to ENOENT for KeyError
+           raise ErrorResponse.wrap(kerr, return_value=-errno.ENOENT)
+        except (BusinessLogcError, OSError) as err:
+           # return value is based on err when possible
+           raise ErrorResponse.wrap(err)
+
+Most uses of ErrorResponse are expected to use the `wrap` classmethod,
+as it will aid in the handling of an existing exception but `ErrorResponse`
+can be used directly too.
+
+== Customizing Response Formatting ==
+
+The `Responder` is built using two additional mid-layer types. The
+`ObjectFormatAdapter` and the `ReturnValueAdapter` by default. These types
+implement the `CommonFormatter` protocol and `ReturnValueProvider` protocols
+respectively. Most cases will not need to customize the `ReturnValueAdapter` as
+returning zero on success is expected.  However, if there's a need to return a
+non-zero error code outside of an exception, you can add the `mgr_return_value`
+function to the returned type of the endpoint function - causing it to meet the
+`ReturnValueProvider` protocol. Whatever integer that function returns will
+then be used in the response tuple.
+
+The `ObjectFormatAdapter` can operate in two modes. By default, any type
+returned from the endpoint function will be checked for a `to_simplified`
+method (the type matches the SimpleDataProvider` protocol) and if it exists
+the method will be called and the result serialized. Example:
+    class CoolStuff:
+       def __init__(self, temperature: int, quantity: int) -> None:
+           self.temperature = temperature
+           self.quantity = quantity
+      def to_simplified(self) -> Dict[str, int]:
+         return {"temp": self.temperature, "qty": self.quantity}
+
+    @CLICommand("command name", perm="r")
+    Responder()
+    def create_something_cool(self) -> CoolStuff:
+       cool_stuff: CoolStuff = self._make_cool_stuff()  # implementation
+       return cool_stuff
+
+To be compatible with many existing types in the ceph mgr codebase one can pass
+`compatible=True` to the `ObjectFormatAdapter`. If the type provides a
+`to_json` and/or `to_yaml` method that returns basic python types (dict, list,
+str, etc...) but *not* already serialized JSON or YAML this flag can be
+enabled. Note that Responder takes as an argument any callable that returns a
+`CommonFormatter`. In this example below we enable the flag using
+`functools.partial`:
+    class MyExistingClass:
+      def to_json(self) -> Dict[str, Any]:
+         return {"name": self.name, "height": self.height}
+
+    @CLICommand("command name", perm="r")
+    Responder(functools.partial(ObjectFormatAdapter, compatible=True))
```

Comment:
it's good that this compatibility functionality is available although it would be nice if there was a more intuitive syntax for it.

Reply:
Agreed. I suggest we wait and see if there's demand for it in the codebase. If we find mgr modules that end up wanting to use it, we can make this more elegant... but I don't want to try to imagine all possible cases for YAGNI reasons.